src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/VisualStudioForegroundDispatcher.cs

```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Composition;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Razor;
+
+namespace Microsoft.VisualStudio.Mac.LanguageServices.Razor
+{
+    [Shared]
+    [ExportWorkspaceService(typeof(ForegroundDispatcher), ServiceLayer.Host)]
+    internal class VisualStudioForegroundDispatcher : ForegroundDispatcher
+    {
+        public override TaskScheduler BackgroundScheduler { get; } = TaskScheduler.Default;
+
+        public override TaskScheduler ForegroundScheduler { get; } = TaskScheduler.FromCurrentSynchronizationContext();
+
+        public override bool IsForegroundThread => MonoDevelop.Core.Runtime.IsMainThread;
```

Comment:
Can't utilize a using here. It expects `Microsoft.Runtime` if I try to.

Reply:
We can't guarantee that the MEF composition will happen on the UI thread.

Use the MainTaskScheduler instead:
http://source.monodevelop.com/#MonoDevelop.Core/MonoDevelop.Core/Runtime.cs,d50b51ce3b3aed98

Reply:
Sorry, this one: http://source.monodevelop.com/#MonoDevelop.Core/MonoDevelop.Core/Runtime.cs,66bf11b517f553a5

Reply:
The MainTaskScheduler doesn't seem to align with the "ForegroundThread" in practice. For example, in the `ITextViewConnectionListener.SubjectBuffersConnected` event handler (expected to be invoked on the foreground thread) this is what I'm seeing:

![image](https://user-images.githubusercontent.com/2008729/33581503-f896c4e8-d905-11e7-9b56-d746c5a7df8b.png)

Reply:
and the full call stack at that point indicates that we're indeed the UI thread? I think there can be more than one TaskSchedulers for the UI Thread. Weird.

You can commit whatever works for you, for now. I should investigate this - it might be an actual bug in MonoDevelop.

Reply:
Filed https://.../DevDiv/_workitems/edit/533116 to track this.